server/src/main/scala/org/broadinstitute/dsp/workbench/welder/server/ObjectService.scala

```diff
@@ -193,13 +215,19 @@ class ObjectService(
         Kleisli.liftF[IO, TraceId, Unit](IO.raiseError(DeleteSafeModeFileError(s"${req.localObjectPath} can't be deleted since it's in safe mode")))
       else {
         val gsPath = getGsPath(req.localObjectPath, context)
-        for {
-          previousMeta <- Kleisli.liftF(metadataCache.get.map(_.get(req.localObjectPath.asPath)))
-          _ <- previousMeta match {
-            case Some(m) => Kleisli.liftF[IO, TraceId, Unit](googleStorageAlg.removeObject(gsPath, traceId, Some(m.generation)).compile.drain.void)
-            case None => Kleisli.liftF[IO, TraceId, Unit](IO.raiseError(UnknownFileState(s"Local GCS metadata for ${req.localObjectPath} not found")))
+        val res = for {
+          previousMeta <- metadataCache.get.map(_.get(req.localObjectPath))
+          now <- timer.clock.realTime(TimeUnit.MILLISECONDS)
+          generation <- previousMeta match {
+            case Some(meta) =>
+              checkLock(meta.lock, now, gsPath.bucketName).as(meta.localFileStateInGCS.map(_.generation).getOrElse(0L)) //check if user owns the lock before deleting
+            case None =>
+              IO.raiseError(InvalidLock(s"Local GCS metadata for ${req.localObjectPath} not found"))
           }
+          _ <- googleStorageAlg.removeObject(gsPath, traceId, Some(generation)).compile.drain.void
```

Comment:
I don't follow the Scala - if this fails do we throw an exception here?

Reply:
yes